private/buf/bufsync/commits_to_sync.go

```diff
@@ -0,0 +1,226 @@
+// Copyright 2020-2023 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bufsync
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/bufbuild/buf/private/bufpkg/bufmodule/bufmodulebuild"
+	"github.com/bufbuild/buf/private/bufpkg/bufmodule/bufmoduleref"
+	"github.com/bufbuild/buf/private/pkg/git"
+	"go.uber.org/zap"
+)
+
+// syncableCommit holds the modules that need to be synced in a git commit.
+type syncableCommit struct {
+	commit  git.Commit
+	modules map[string]bufmodulebuild.BuiltModule // moduleDir:builtModule
+}
+
+// branchCommitsToSync returns a sorted commit+modules tuples array that are pending to sync for a
+// branch. A commit in the array might have no modules to sync if those are skipped by the
+// Syncer error handler, or are a found sync point.
+func (s *syncer) branchCommitsToSync(ctx context.Context, branch string) ([]syncableCommit, error) {
+	modulesToSync, ok := s.branchesModulesToSync[branch]
+	if !ok || len(modulesToSync) == 0 {
+		// branch should not be synced, or no modules to sync in that branch
+		return nil, nil
+	}
+	// Copy all branch modules to sync and mark them as pending, until its starting sync point is
+	// reached. They'll be removed from this list as its initial sync point is found.
+	pendingModules := s.copyBranchModulesSync(branch, modulesToSync)
+	var commitsToSync []syncableCommit
+	// travel branch commits from HEAD and check if they're already synced, until finding a synced git
+	// commit, or adding them all to be synced
+	stopLoopErr := errors.New("stop loop")
+	if err := s.repo.ForEachCommit(branch, func(commit git.Commit) error {
+		if len(pendingModules) == 0 {
+			// no more pending modules to sync, no need to keep navigating the branch
+			return stopLoopErr
+		}
+		commitHash := commit.Hash().Hex()
+		modulesDirsToSyncInThisCommit := make(map[string]bufmodulebuild.BuiltModule)
```

Comment:
We will hold the built module in here, so we read each module on each commit just once, instead of reading it here and again while syncing.